pkg/client/events/replay_client.go

```diff
@@ -0,0 +1,244 @@
+	"github.com/pokt-network/poktroll/pkg/observable/channel"
+	"github.com/pokt-network/poktroll/pkg/retry"
+)
+
+const (
+	// eventsBytesRetryDelay is the delay between retry attempts when the events
+	// bytes observable returns an error.
+	eventsBytesRetryDelay = time.Second
+	// eventsBytesRetryLimit is the maximum number of times to attempt to
+	// re-establish the events query bytes subscription when the events bytes
+	// observable returns an error.
+	eventsBytesRetryLimit        = 10
+	eventsBytesRetryResetTimeout = 10 * time.Second
+	// latestObsvblsReplayBufferSize is the replay buffer size of the
+	// latestObsvbls replay observable which is used to cache the latest observable.
+	// It is updated with a new "active" observable when a new
+	// events query subscription is created, for example, after a non-persistent
+	// connection error.
+	latestObsvblsReplayBufferSize = 1
+	// latestReplayBufferSize is the replay buffer size of the latest replay
+	// observable which is notified when block commit events are received
+	// by the events query client subscription created in goPublishEvents.
+	latestReplayBufferSize = 1
+)
+
+// Enforece the EventsReplayClient interface is implemented by the replayClient type.
+var _ client.EventsReplayClient[
+	any,
+	observable.ReplayObservable[any],
+] = (*replayClient[any, observable.ReplayObservable[any]])(nil)
+
+// replayClient implements the EventsReplayClient interface for a generic type T,
+// and replay observable for type T.
+type replayClient[T any, U observable.ReplayObservable[T]] struct {
+	// endpointURL is the URL of RPC endpoint which eventsClient subscription
+	// requests will be sent.
+	endpointURL string
+	// queryString is the query string used to subscribe to events of the
+	// desired type.
+	// See: https://docs.cosmos.network/main/learn/advanced/events#subscribing-to-events
+	// and: https://docs.cosmos.network/main/learn/advanced/events#default-events
+	queryString string
+	// eventsClient is the events query client which is used to subscribe to
+	// newly committed block events. It emits an either value which may contain
+	// an error, at most, once and closes immediately after if it does.
+	eventsClient client.EventsQueryClient
+	// eventBytesToTypeDecoder is a function which decodes event subscription
+	// message bytes into the type defined by the EventsReplayClient's generic type
+	// parameter.
+	eventBytesToTypeDecoder func([]byte) (T, error)
+	// latestObsvbls is a replay observable with replay buffer size 1,
+	// which holds the "active latest observable" which is notified when
+	// new events are received by the events query client subscription
+	// created in goPublishEvents. This observable (and the one it emits) closes
+	// when the events bytes observable returns an error and is updated with a
+	// new "active" observable after a new events query subscription is created.
+	latestObsvbls observable.ReplayObservable[U]
```

Comment:
Wdyt about renaming this to `latestObs`. Open to better suggestions but I think the consistency would be an improvement.

Reply:
(cc @Olshansk)

Reply:
See https://github.com/pokt-network/poktroll/pull/220#discussion_r1415968048